File: tsd/core/types/marketdata.py
from datetime import datetime
from enum import StrEnum
from typing import List

import csp

from core.types.exchange import Exchange
from core.types.instrument import Instrument


class Side(StrEnum):
    BID = "BID"
    ASK = "ASK"


class TwoWayPrice(csp.Struct):
    """
    Represents a snapshot of the best bid and ask prices and quantities
    for a given instrument at a specific time.

    Attributes:
        bid_price (float): Best bid price available.
        ask_price (float): Best ask price available.
        bid_qty (float): Quantity available at the best bid price.
        ask_qty (float): Quantity available at the best ask price.
        time_exchange (int): Time (in nanos) when the quote was recorded at the exchange.
        time_received (int): Time (in nanos) when the quote was received by the system.
    """
    bid_price: float
    ask_price: float
    bid_qty: float
    ask_qty: float
    time_exchange: int
    time_received: int


class MarketOrder(csp.Struct):
    """
    Represents a single market order received from an exchange.

    Attributes:
        instr (Instrument): Instrument the order relates to.
        exchange (Exchange): Exchange from which the order originated.
        side (Side): Indicates whether the order is a bid or an ask.
        price (float): Price level of the order.
        qty (float): Quantity specified in the order.
        time_exchange (int): Time (in nanos) when the order was registered at the exchange.
        time_received (int): Time (in nanos) when the order was received by the system.
    """
    instr: Instrument
    exchange: Exchange
    side: Side
    price: float
    qty: float
    time_exchange: datetime
    time_received: datetime

    def __repr__(self):
        return f"MarketOrder( instr={self.instr.symbol}, exchange={self.exchange.value}, side={self.side}, price={self.price}, qty={self.qty}, time_exchange={self.time_exchange}, time_received={self.time_received} )"

    def __str__(self):
        return f"MarketOrder( instr={self.instr.symbol}, exchange={self.exchange.value}, side={self.side}, price={self.price}, qty={self.qty}, time_exchange={self.time_exchange}, time_received={self.time_received} )"


class OrderBook(csp.Struct):
    """
    Represents the full depth of market for a given instrument at a specific time,
    containing both bid and ask orders.

    Attributes:
        instr (Instrument): Instrument to which this order book pertains.
        bids (List[MarketOrder]): List of bid-side orders sorted by price and/or time priority.
        asks (List[MarketOrder]): List of ask-side orders sorted by price and/or time priority.
        time_exchange (int): Time (in nanos) when the order book snapshot was recorded at the exchange.
        time_received (int): Time (in nanos) when the snapshot was received by the system.
    """
    instr: Instrument
    bids: List[MarketOrder]
    asks: List[MarketOrder]
    time_exchange: datetime
    time_received: datetime

    def __repr__(self):
        return f"OrderBook( instr={self.instr.symbol}, bids={self.bids}, asks={self.asks}, time_exchange={self.time_exchange}, time_received={self.time_received} )"

    def __str__(self):
        return f"OrderBook( instr={self.instr.symbol}, bids={self.bids}, asks={self.asks}, time_exchange={self.time_exchange}, time_received={self.time_received} )"

File: tsd/tsd_cpp/core/types/marketdata.cpp
#include "marketdata.hpp"
#include <sstream>
#include <algorithm>
#include <iomanip>

namespace tsd {
namespace core {

std::string to_string(Side side) {
    switch (side) {
        case Side::BID: return "BID";
        case Side::ASK: return "ASK";
        default: return "UNKNOWN";
    }
}

Side side_from_string(const std::string& str) {
    if (str == "BID") return Side::BID;
    if (str == "ASK") return Side::ASK;
    throw std::invalid_argument("Unknown Side: " + str);
}

std::string MarketOrder::to_string() const {
    std::stringstream ss;
    ss << "MarketOrder( instr=" << instr.symbol()
       << ", exchange=" << tsd::core::to_string(exchange)
       << ", side=" << tsd::core::to_string(side)
       << ", price=" << std::fixed << std::setprecision(8) << price
       << ", qty=" << std::fixed << std::setprecision(8) << qty
       << ", time_exchange=" << std::chrono::duration_cast<std::chrono::milliseconds>(time_exchange.time_since_epoch()).count() << "ms"
       << ", time_received=" << std::chrono::duration_cast<std::chrono::milliseconds>(time_received.time_since_epoch()).count() << "ms )";
    return ss.str();
}

double OrderBook::best_bid() const {
    if (bids.empty()) return 0.0;

    auto max_bid = std::max_element(bids.begin(), bids.end(),
        [](const MarketOrder& a, const MarketOrder& b) {
            return a.price < b.price;
        });

    return max_bid->price;
}

double OrderBook::best_ask() const {
    if (asks.empty()) return 0.0;

    auto min_ask = std::min_element(asks.begin(), asks.end(),
        [](const MarketOrder& a, const MarketOrder& b) {
            return a.price < b.price;
        });

    return min_ask->price;
}

double OrderBook::spread() const {
    double bid = best_bid();
    double ask = best_ask();
    return (bid > 0 && ask > 0) ? ask - bid : 0.0;
}

double OrderBook::mid_price() const {
    double bid = best_bid();
    double ask = best_ask();
    return (bid > 0 && ask > 0) ? (bid + ask) / 2.0 : 0.0;
}

std::string OrderBook::to_string() const {
    std::stringstream ss;
    ss << "OrderBook( instr=" << instr.symbol()
       << ", bids=[";

    for (size_t i = 0; i < bids.size() && i < 3; ++i) { // Show first 3 levels
        if (i > 0) ss << ", ";
        ss << std::fixed << std::setprecision(4) << bids[i].price << "@" << bids[i].qty;
    }

    ss << "], asks=[";

    for (size_t i = 0; i < asks.size() && i < 3; ++i) { // Show first 3 levels
        if (i > 0) ss << ", ";
        ss << std::fixed << std::setprecision(4) << asks[i].price << "@" << asks[i].qty;
    }

    ss << "], time_exchange=" << std::chrono::duration_cast<std::chrono::milliseconds>(time_exchange.time_since_epoch()).count() << "ms"
       << ", time_received=" << std::chrono::duration_cast<std::chrono::milliseconds>(time_received.time_since_epoch()).count() << "ms )";

    return ss.str();
}

std::ostream& operator<<(std::ostream& os, const Side& side) {
    return os << to_string(side);
}

std::ostream& operator<<(std::ostream& os, const TwoWayPrice& price) {
    return os << "TwoWayPrice( bid=" << std::fixed << std::setprecision(4) << price.bid_price
              << "@" << price.bid_qty << ", ask=" << price.ask_price << "@" << price.ask_qty << " )";
}

std::ostream& operator<<(std::ostream& os, const MarketOrder& order) {
    return os << order.to_string();
}

std::ostream& operator<<(std::ostream& os, const OrderBook& book) {
    return os << book.to_string();
}

} // namespace core
} // namespace tsd